File: app/src/main/java/com/example/mutidemo/ui/FacePreViewActivity.java
package com.example.mutidemo.ui;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.ImageFormat;
import android.graphics.Matrix;
import android.graphics.Rect;
import android.hardware.Camera;
import android.media.FaceDetector;
import android.util.Log;
import android.view.SurfaceHolder;

import androidx.annotation.NonNull;

import com.example.mutidemo.databinding.ActivityFaceBinding;
import com.pengxh.androidx.lite.base.AndroidxBaseActivity;
import com.pengxh.androidx.lite.utils.ImageUtil;

import java.io.IOException;
import java.util.List;
import java.util.Stack;

public class FacePreViewActivity extends AndroidxBaseActivity<ActivityFaceBinding> implements Camera.PreviewCallback {

    private static final String TAG = "FacePreViewActivity";
    private SurfaceHolder mSurfaceHolder;
    private Camera mCamera;
    private Stack<Bitmap> bitmapStack;

    @Override
    protected void setupTopBarLayout() {

    }

    @Override
    public void initData() {
        bitmapStack = new Stack<>();
    }

    @Override
    public void initEvent() {
        // 绑定SurfaceView，取得SurfaceHolder对象
        mSurfaceHolder = viewBinding.surfaceView.getHolder();
        mSurfaceHolder.addCallback(new SurfaceHolder.Callback() {
            @Override
            public void surfaceCreated(@NonNull SurfaceHolder holder) {
                if (mCamera == null) {
                    //打开相机
                    openCamera();
                }
                try {
                    //预览画面
                    mCamera.setPreviewDisplay(mSurfaceHolder);
                } catch (IOException e) {
                    e.printStackTrace();
                }
                //开始预览
                mCamera.startPreview();
                //人脸检测
                mCamera.startFaceDetection();
                mCamera.setFaceDetectionListener(new Camera.FaceDetectionListener() {
                    @Override
                    public void onFaceDetection(Camera.Face[] faces, Camera camera) {
                        if (faces.length > 0) {
                            Camera.Face face = faces[0];
                            Rect rect = face.rect;
                            Log.d(TAG, "可信度：" + face.score +
                                    " ,face detected: " + faces.length +
                                    " ,X: " + rect.centerX() +
                                    " ,Y: " + rect.centerY() +
                                    " ,[" + rect.left + "," + rect.top + "," + rect.right + "," + rect.bottom + "]");
                            Matrix matrix = updateFaceRect();
                            viewBinding.faceDetectView.updateFace(matrix, faces);
                            viewBinding.faceTipsView.setText("已检测到人脸，识别中");
                            viewBinding.faceTipsView.setTextColor(Color.GREEN);
                        } else {
                            viewBinding.faceDetectView.removeRect();
                        }
                    }
                });
            }

            @Override
            public void surfaceChanged(@NonNull SurfaceHolder holder, int format, int width, int height) {

            }

            @Override
            public void surfaceDestroyed(@NonNull SurfaceHolder holder) {
                releaseCamera(); //释放相机资源
            }
        });
        mSurfaceHolder.setType(SurfaceHolder.SURFACE_TYPE_PUSH_BUFFERS);// setType必须设置
    }

    private Matrix updateFaceRect() {
        Matrix matrix = new Matrix();
        Camera.CameraInfo info = new android.hardware.Camera.CameraInfo();
        // Need mirror for front camera.
        boolean mirror = (info.facing == Camera.CameraInfo.CAMERA_FACING_FRONT);
        matrix.setScale(mirror ? -1 : 1, 1);
        // This is the value for android.hardware.Camera.setDisplayOrientation.
        // 刚才我们设置了camera的旋转参数，所以这里也要设置一下
        matrix.postRotate(90);
        // Camera driver coordinates range from (-1000, -1000) to (1000, 1000).
        // UI coordinates range from (0, 0) to (width, height).
        matrix.postScale(viewBinding.surfaceView.getWidth() / 2000f, viewBinding.surfaceView.getHeight() / 2000f);
        matrix.postTranslate(viewBinding.surfaceView.getWidth() / 2f, viewBinding.surfaceView.getHeight() / 2f);
        return matrix;
    }

    @Override
    public void onPreviewFrame(byte[] data, Camera camera) {
        camera.addCallbackBuffer(data);
        Camera.Size size = camera.getParameters().getPreviewSize();//必须是相机支持的预览尺寸，否则颜色YUV空间会错位
        Bitmap originBitmap = ImageUtil.nv21ToBitmap(data, size.width, size.height);
        //要使用Android内置的人脸识别，需要将Bitmap对象转为RGB_565格式，否则无法识别
        Bitmap faceDetectorBitmap = originBitmap.copy(Bitmap.Config.RGB_565, true);
        FaceDetector.Face[] faces = new FaceDetector.Face[1];
        FaceDetector faceDetector = new FaceDetector(faceDetectorBitmap.getWidth(), faceDetectorBitmap.getHeight(), 1);
        int faceSum = faceDetector.findFaces(faceDetectorBitmap, faces);
        if (faceSum == 1) {
            bitmapStack.push(originBitmap);
            if (bitmapStack.size() >= 3) {//当栈里有3张bitmap之后才开始识别
                Bitmap bitmap = bitmapStack.pop();
                Intent intent = new Intent();
                intent.putExtra("imageToBase64", ImageUtil.imageToBase64(bitmap));
                setResult(Activity.RESULT_OK, intent);
                finish();
            }
        }
    }

    /**
     * 打开相机
     */
    private void openCamera() {
        try {
            mCamera = Camera.open(1);
            initParameters(mCamera);        //初始化相机配置信息
            mCamera.setPreviewCallback(this);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 初始化相机属性
     */
    private void initParameters(Camera camera) {
        try {
            Camera.Parameters mParameters = camera.getParameters();
            mParameters.setPreviewFormat(ImageFormat.NV21);  //设置预览图片的格式
            //获取与指定宽高相等或最接近的尺寸
            //设置预览尺寸
            Camera.Size bestPreviewSize = obtainBestSize(viewBinding.surfaceView.getWidth(), viewBinding.surfaceView.getHeight(), mParameters.getSupportedPreviewSizes());
            mParameters.setPreviewSize(bestPreviewSize.width, bestPreviewSize.height);
            //设置保存图片尺寸
//            Camera.Size bestPicSize = obtainBestSize(picWidth, picHeight, mParameters.getSupportedPictureSizes());
//            mParameters.setPictureSize(bestPicSize.width, bestPicSize.height);
            //对焦模式
            mParameters.setFocusMode(Camera.Parameters.FOCUS_MODE_CONTINUOUS_PICTURE);
            camera.setDisplayOrientation(90);
            camera.setParameters(mParameters);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 获取最合适的尺寸
     */
    private Camera.Size obtainBestSize(int targetWidth, int targetHeight, List<Camera.Size> sizeList) {
        Camera.Size bestSize = null;
        int targetRatio = targetHeight / targetWidth;//目标大小的宽高比
        int minDiff = targetRatio;

        for (Camera.Size size : sizeList) {
            if (size.width == targetHeight && size.height == targetWidth) {
                bestSize = size;
                break;
            }
            int supportedRatio = (size.width / size.height);
            if (Math.abs(supportedRatio - targetRatio) < minDiff) {
                minDiff = Math.abs(supportedRatio - targetRatio);
                bestSize = size;
            }
        }
        return bestSize;
    }

    /**
     * 释放相机资源
     */
    private void releaseCamera() {
        if (mCamera != null) {
            mCamera.stopPreview();
            mCamera.setPreviewCallback(null);
            mCamera.release();
            mCamera = null;
        }
    }
}

File: app/src/main/java/com/example/mutidemo/view/FacePreViewActivity.kt
package com.example.mutidemo.view

import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.ImageFormat
import android.graphics.Matrix
import android.hardware.Camera
import android.media.FaceDetector
import android.util.Log
import android.view.SurfaceHolder
import com.example.mutidemo.R
import com.pengxh.kt.lite.base.KotlinBaseActivity
import com.pengxh.kt.lite.extensions.nv21ToBitmap
import com.pengxh.kt.lite.extensions.toBase64
import kotlinx.android.synthetic.main.activity_face.*
import java.io.IOException
import java.util.*
import kotlin.math.abs

class FacePreViewActivity : KotlinBaseActivity(), Camera.PreviewCallback {

    private val kTag = "FacePreViewActivity"
    private val bitmapStack by lazy { Stack<Bitmap>() }
    private lateinit var surfaceHolder: SurfaceHolder
    private var camera: Camera? = null

    override fun observeRequestState() {

    }

    override fun initLayoutView(): Int = R.layout.activity_face

    override fun setupTopBarLayout() {}

    override fun initData() {
        // 绑定SurfaceView，取得SurfaceHolder对象
        surfaceHolder = surfaceView.holder
        surfaceHolder.addCallback(object : SurfaceHolder.Callback {
            override fun surfaceCreated(holder: SurfaceHolder) {
                if (camera == null) {
                    //打开相机
                    openCamera()
                }
                try {
                    //预览画面
                    camera?.setPreviewDisplay(surfaceHolder)
                } catch (e: IOException) {
                    e.printStackTrace()
                }
                //开始预览
                camera?.startPreview()
                //人脸检测
                camera?.startFaceDetection()
                camera?.setFaceDetectionListener { faces, _ ->
                    if (faces.isNotEmpty()) {
                        val face = faces[0]
                        val rect = face.rect
                        Log.d(
                            kTag, "可信度：" + face.score +
                                    " ,face detected: " + faces.size +
                                    " ,X: " + rect.centerX() +
                                    " ,Y: " + rect.centerY() +
                                    " ,[" + rect.left + "," + rect.top + "," + rect.right + "," + rect.bottom + "]"
                        )
                        val matrix = updateFaceRect()
                        faceDetectView.updateFace(matrix, faces)
                        faceTipsView.text = "已检测到人脸，识别中"
                        faceTipsView.setTextColor(Color.GREEN)
                    } else {
                        faceDetectView.removeRect()
                    }
                }
            }

            override fun surfaceChanged(
                holder: SurfaceHolder, format: Int, width: Int, height: Int
            ) {
            }

            override fun surfaceDestroyed(holder: SurfaceHolder) {
                releaseCamera() //释放相机资源
            }
        })
        surfaceHolder.setType(SurfaceHolder.SURFACE_TYPE_PUSH_BUFFERS) // setType必须设置
    }

    /**
     * 打开相机
     */
    private fun openCamera() {
        try {
            camera = Camera.open(1)
            initParameters(camera) //初始化相机配置信息
            camera?.setPreviewCallback(this)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 初始化相机属性
     */
    private fun initParameters(camera: Camera?) {
        try {
            val cameraParameters = camera?.parameters ?: return
            cameraParameters.previewFormat = ImageFormat.NV21 //设置预览图片的格式
            //获取与指定宽高相等或最接近的尺寸
            //设置预览尺寸
            val bestPreviewSize = obtainBestSize(
                surfaceView.width,
                surfaceView.height,
                cameraParameters.supportedPreviewSizes
            )
            cameraParameters.setPreviewSize(bestPreviewSize!!.width, bestPreviewSize.height)
            //设置保存图片尺寸
//            Camera.Size bestPicSize = obtainBestSize(picWidth, picHeight, cameraParameters.getSupportedPictureSizes());
//            cameraParameters.setPictureSize(bestPicSize.width, bestPicSize.height);
            //对焦模式
            cameraParameters.focusMode = Camera.Parameters.FOCUS_MODE_CONTINUOUS_PICTURE
            camera.setDisplayOrientation(90)
            camera.parameters = cameraParameters
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 获取最合适的尺寸
     */
    private fun obtainBestSize(
        targetWidth: Int, targetHeight: Int, sizeList: List<Camera.Size>
    ): Camera.Size? {
        var bestSize: Camera.Size? = null
        val targetRatio = targetHeight / targetWidth //目标大小的宽高比
        var minDiff = targetRatio
        for (size in sizeList) {
            if (size.width == targetHeight && size.height == targetWidth) {
                bestSize = size
                break
            }
            val supportedRatio = size.width / size.height
            if (abs(supportedRatio - targetRatio) < minDiff) {
                minDiff = abs(supportedRatio - targetRatio)
                bestSize = size
            }
        }
        return bestSize
    }

    override fun initEvent() {

    }

    private fun updateFaceRect(): Matrix {
        val matrix = Matrix()
        val info = Camera.CameraInfo()
        // Need mirror for front camera.
        val mirror = info.facing == Camera.CameraInfo.CAMERA_FACING_FRONT
        val sx = if (mirror) {
            -1f
        } else {
            1f
        }
        matrix.setScale(sx, 1f)
        // This is the value for android.hardware.Camera.setDisplayOrientation.
        // 刚才我们设置了camera的旋转参数，所以这里也要设置一下
        matrix.postRotate(90f)
        // Camera driver coordinates range from (-1000, -1000) to (1000, 1000).
        // UI coordinates range from (0, 0) to (width, height).
        matrix.postScale(
            surfaceView.width / 2000f,
            surfaceView.height / 2000f
        )
        matrix.postTranslate(
            surfaceView.width / 2f,
            surfaceView.height / 2f
        )
        return matrix
    }

    override fun onPreviewFrame(data: ByteArray, camera: Camera) {
        camera.addCallbackBuffer(data)
        val size = camera.parameters.previewSize //必须是相机支持的预览尺寸，否则颜色YUV空间会错位
        val originBitmap = data.nv21ToBitmap(size.width, size.height)
        //要使用Android内置的人脸识别，需要将Bitmap对象转为RGB_565格式，否则无法识别
        val faceDetectorBitmap: Bitmap = originBitmap!!.copy(Bitmap.Config.RGB_565, true)
        val faces = arrayOfNulls<FaceDetector.Face>(1)
        val faceDetector =
            FaceDetector(faceDetectorBitmap.width, faceDetectorBitmap.height, 1)
        val faceSum: Int = faceDetector.findFaces(faceDetectorBitmap, faces)
        if (faceSum == 1) {
            bitmapStack.push(originBitmap)
            if (bitmapStack.size >= 3) { //当栈里有3张bitmap之后才开始识别
                val bitmap: Bitmap = bitmapStack.pop()
                val intent = Intent()
                intent.putExtra("imageToBase64", bitmap.toBase64())
                setResult(Activity.RESULT_OK, intent)
                finish()
            }
        }
    }

    /**
     * 释放相机资源
     */
    private fun releaseCamera() {
        camera?.stopPreview()
        camera?.setPreviewCallback(null)
        camera?.release()
        camera = null
    }
}